modules/OSSMail/models/Mail.php

```diff
@@ -253,65 +252,115 @@ public function getEmail($name)
 		return $return;
 	}
 
+	/**
+	 * Search crmids by emails.
+	 *
+	 * @param string[] $emails
+	 * @param string   $moduleName
+	 * @param string   $fieldName
+	 *
+	 * @return array crmids
+	 */
+	public function searchByEmails(array $emails, string $moduleName, string $fieldName)
+	{
+		$return = [];
+		foreach ($emails as $email) {
+			if (empty($email)) {
+				continue;
+			}
+			$name = 'MSFindEmail_' . $moduleName . '_' . $fieldName;
+			$cache = App\Cache::get($name, $email);
+			if ($cache !== false) {
+				if ($cache != 0) {
+					$return = array_merge($return, $cache);
+				}
+			} else {
+				$ids = [];
+				$queryGenerator = new \App\QueryGenerator($moduleName);
+				if ($queryGenerator->getModuleField($fieldName)) {
+					$queryGenerator->setFields(['id']);
+					$queryGenerator->addCondition($fieldName, $email, 'e');
+					$dataReader = $queryGenerator->createQuery()->createCommand()->query();
+					while (($crmid = $dataReader->readColumn(0)) !== false) {
+						$ids[] = $crmid;
+					}
+					$dataReader->close();
+					$return = array_merge($return, $ids);
+				}
+				if (empty($ids)) {
+					$ids = 0;
+				}
+				App\Cache::save($name, $email, $ids);
+			}
+		}
+		return $return;
+	}
+
+	/**
+	 * Search crmids from domains.
+	 *
+	 * @param Vtiger_Field_Model $fieldModel
+	 * @param string[]           $emails
+	 *
+	 * @return array crmids
+	 */
+	public function searchByDomains(Vtiger_Field_Model $fieldModel, array $emails)
+	{
+		$fieldName = $fieldModel->getName();
+		$moduleName = $fieldModel->getModuleName();
+		$name = 'MSFindEmail_' . $moduleName . '_' . $fieldName;
+		$crmids = [];
+		foreach ($emails as $email) {
+			$domain = mb_strtolower(explode('@', $email)[1]);
+			$cache = App\Cache::get($name, $domain);
+			if ($cache !== false) {
+				if ($cache != 0) {
+					$crmids = array_merge($crmids, $cache);
+				}
+			} else {
+				$crmids = App\Fields\MultiDomain::getCrmIds($domain, $fieldModel);
+				App\Cache::save($name, $domain, $crmids);
+			}
+		}
+		return $crmids;
+	}
+
 	/**
 	 * Find email address.
 	 *
 	 * @param string $field
 	 * @param string $searchModule
 	 * @param bool   $returnArray
 	 *
-	 * @return string|array
+	 * @return array|string
 	 */
 	public function findEmailAdress($field, $searchModule = false, $returnArray = true)
 	{
 		$return = [];
 		$emails = $this->get($field);
-		$emailSearchList = OSSMailScanner_Record_Model::getEmailSearchList();
-
 		if (empty($emails)) {
 			return [];
-		} elseif (strpos($emails, ',')) {
+		}
+		if (strpos($emails, ',')) {
 			$emails = explode(',', $emails);
 		} else {
 			$emails = (array) $emails;
 		}
+		$emailSearchList = OSSMailScanner_Record_Model::getEmailSearchList();
 		if (!empty($emailSearchList)) {
 			foreach ($emailSearchList as $field) {
 				$enableFind = true;
 				$row = explode('=', $field);
 				$moduleName = $row[1];
+				$fieldModel = Vtiger_Field_Model::getInstance($row[0], Vtiger_Module_Model::getInstance($moduleName));
```

Comment:
Vtiger_Module_Model::getInstance($moduleName)  
before the loop

Reply:
$moduleName is specified inside loop